src/dotnet-setversion/Program.cs

```diff
@@ -2,44 +2,107 @@
 using System.IO;
 using System.Linq;
 using System.Xml.Linq;
+using CommandLine;
 
 namespace dotnet_setversion
 {
-    public static class Program
+    internal static class Program
     {
-        private static int Main(string[] args)
+        internal static int Main(params string[] args)
         {
-            if (args.Length < 1)
+            return Parser.Default.ParseArguments<Options>(args)
+                .MapResult(Run, errs => int.MinValue);
+        }
+
+        private static int Run(Options options)
+        {
+            if (options.Recursive && options.CsprojFile != null)
             {
-                Console.WriteLine("Missing version string");
-                return 1;
+                Console.WriteLine("The ---recurvice option and csprojFile argument are mutually exclusive.");
+                return -1;
             }
 
-            var versionString = args[0];
-            var currentDirectory = Directory.GetCurrentDirectory();
-            var csprojFile = Directory.EnumerateFileSystemEntries(currentDirectory, "*.csproj", SearchOption.TopDirectoryOnly).First();
-            var document = XDocument.Load(csprojFile);
+            return options.Recursive ? RunRecursive(options.Version) : Run(options.Version, options.CsprojFile);
+        }
 
-            // Search for an existing version node in the csproj file.
-            var projectNode = document.GetOrCreateElement("Project");
-            var versionNode = projectNode
-                .Elements("PropertyGroup")
-                .SelectMany(it => it.Elements("Version"))
-                .SingleOrDefault();
+        private static int Run(string version, string csprojFile)
+        {
+            if (csprojFile == null)
+            {
+                var csprojFiles = GetCsprojFiles(false);
+                if (!CheckCsprojFiles(csprojFiles, false, out var exitCode)) return exitCode;
+                csprojFile = csprojFiles[0];
+            }
 
-            // If no version node exists, create it.
-            if (versionNode == null)
+            if (!File.Exists(csprojFile))
             {
-                versionNode = projectNode
-                    .GetOrCreateElement("PropertyGroup")
-                    .GetOrCreateElement("Version");
+                Console.WriteLine($"Project file '{csprojFile}' does not exist.");
+                return -1009;
```

Comment:
What is the significance of -1009 and other error codes you're using?

Reply:
These codes correspond to the MSBuild error codes [MSB1003](https://msdn.microsoft.com/en-us/library/w9615yb2.aspx), [MSB1009](https://msdn.microsoft.com/en-us/library/fbzcadyb.aspx) and [MSB1011](https://msdn.microsoft.com/en-us/library/8wyxzk67.aspx)

Reply:
Ah fair, but `dotnet build`, `dotnet restore` etc. return error code 1 regardless of the error code they report. If we want to maintain consistency with those tools, we'd return 0 for success and 1 for any sort of failure.

Reply:
Agreed.